tooling/lsp/src/lib.rs

```diff
@@ -175,3 +189,29 @@ fn byte_span_to_range<'a, F: files::Files<'a> + ?Sized>(
         None
     }
 }
+
+pub(crate) fn resolve_workspace_for_source_path(
+    file_path: &Path,
+) -> Result<nargo::workspace::Workspace, LspError> {
+    let package_root = find_file_manifest(file_path);
+
+    let toml_path = match package_root {
+        Some(toml_path) => toml_path,
+        None => {
+            return Err(LspError::WorkspaceResolutionError(format!(
+                "Nargo.toml not found for file: {:?}",
+                file_path
+            )))
+        }
+    };
+
+    let workspace = match resolve_workspace_from_toml(
```

Comment:
you can also avoid the match and return $resolve_workspace_from_toml(..)$ and convert the error at the same time. (I think).